scigym/envs/surfacecode/surfacecode_env.py

```diff
@@ -0,0 +1,717 @@
+
+		# 2) Calculate the reward
+		self.current_true_syndrome = self.generate_surface_code_syndrome_NoFT_efficient(self.hidden_state, self.qubits)
+		current_true_syndrome_vector = np.reshape(self.current_true_syndrome,(self.d+1)**2) 
+		num_anyons = np.sum(self.current_true_syndrome)
+
+		correct_label = self.generate_one_hot_labels_surface_code(self.hidden_state, self.error_model)
+		decoder_label = self.static_decoder.predict(np.array([current_true_syndrome_vector]), batch_size=1, verbose=0)
+
+		reward = 0
+
+		if np.argmax(correct_label) == 0 and num_anyons == 0:
+			reward = 1.0
+		elif np.argmax(decoder_label[0]) != np.argmax(correct_label):
+			self.done = True
+
+
+		# 3) If necessary, apply multiple errors and obtain an error volume - ensure that a non-trivial volume is generated
+		if done_identity:
+
+			trivial_volume = True
+			while trivial_volume: 
+				self.summed_syndrome_volume = np.zeros((self.d + 1, self.d + 1), int)
+				faulty_syndromes = []
+				for j in range(self.volume_depth):
+					error = self.generate_error(self.d, self.p_phys, self.error_model)
+					if int(np.sum(error)!=0):
+						self.hidden_state = self.obtain_new_error_configuration(self.hidden_state, error)
+						self.current_true_syndrome = self.generate_surface_code_syndrome_NoFT_efficient(self.hidden_state, self.qubits)
+					current_faulty_syndrome = self.generate_faulty_syndrome(self.current_true_syndrome, self.p_meas)
+					faulty_syndromes.append(current_faulty_syndrome)
+					self.summed_syndrome_volume += current_faulty_syndrome
+					self.lifetime += 1
+
+				if int(np.sum(self.summed_syndrome_volume)) != 0:
+					trivial_volume = False
+
+			for j in range(self.volume_depth):
+				self.board_state[j, :, :] = self.padding_syndrome(faulty_syndromes[j])
+
+
+			# reset the legal moves
+			self.reset_legal_moves()
+
+			# update the part of the state which shows the actions you have just taken
+			self.board_state[self.volume_depth:,:,:] = np.zeros((self.n_action_layers, 2 * self.d + 1, 2 * self.d + 1),int)
+
+
+		else:
+			# Update the completed actions and legal moves
+			self.completed_actions[action] = int(not(self.completed_actions[action]))
+			if not action == self.identity_index:
+
+				acted_qubit = action%(self.d**2)
+				
+				if acted_qubit not in self.acted_on_qubits:
+					self.acted_on_qubits.add(acted_qubit)
+					for neighbour in self.qubit_neighbours[acted_qubit]:
+							for j in range(self.n_action_layers):
+								self.legal_actions.add(neighbour + j*self.d**2)
+
+				
+			# update the board state to reflect the action thats been taken
+			for k in range(self.n_action_layers):
+					self.board_state[self.volume_depth + k, :, :] = self.padding_actions(self.completed_actions[k * self.d ** 2:(k + 1) * self.d ** 2])
+
+
+		return self.board_state, reward, self.done, {}
+	
+
+	def render(self):
+		raise NotImplementedError
+
+	# ----------------- helper methods ---------------------------------------------------------------------
```

Comment:
This is a formal comment, and you may ignore it if you want: If your `helper functions` are private to the internal workings of your environment, you might want to start them with an underscore, e.g. `def _initialize_state(self): ...`.

Reply:
I think I would prefer to leave these methods without a leading underscore. As the environment is relatively complicated, it may be that users (who decide to look at the source code) want to call helper functions to get a better sense of what is going on within the "black box". From PEP-8:

> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose name starts with an underscore.

As a result, if the leading underscores are added, it might make it _slightly_ more inconvenient to explore what these functions are doing.

Reply:
I would have said that `from env import *` imports all the relevant functionalities for the environment to work, nothing more. In particular, that would not include helper functions. However, I see your point. At the same time, people who want to contribute to your environment are actually going to be thankful for any extra code structuring. If you explicitly distinguish "helper" functions from other, more important functions, that will just help them understand the code. I doubt that anyone will be discouraged to contribute if you give your code some more structure.

Reply:
Ok, I think it is a matter of fixing style, and I'm happy with saying that `from env import * ` should only import methods necessary to use the environment as an API (i.e. everything the agent might need). 

In the end, you are right that more code structure can only help, and one can easily call/import the functions explicitly anyway. So, I'll add leading underscores to all helper functions (defined as methods not _required_ by the `gym.Env` class).